decision(issues): add draft decision about how to create issues and prs (#242)
canonical/observability: decision-records/2024-12-03--creating-issues-in-github-and-jira.md

# Creating issues in GitHub and Jira

**Date:** 2024-12-03<br/>
**Authors:** @simskij

## Context and Problem Statement

As the team have grown, we've ended up diverging in our practices on how to create issues in our
repositories. To be able to work efficiently and consistently with our issues, we need to follow some
common practices, so that the workflow remains the same across issues.

## Decision

These rules only apply to tickets opened by us. If someone else opens the ticket, we should
rephrase it as part of the triage.

### Issue Titles

- Issues are always to be created in GitHub, unless confidentiality requirements dictate something
  else. However, this is almost never the case. If you are unsure - ask @simskij or @sed-i.
- Issues of the story/enhancement type should have a title that describes the work to be done in
  the _imperative modus_, i.e "Add a bip to the bop", rather than "We should add a bip to the bop"
- Issues of the bug type should have a title that describes the bug occuring, written in

>   _modus ponens_, i.e "When <condition> then <outcome>".

Reviewer: ```suggestion
  _modus ponens_, i.e "When <condition> then <outcome>". When necessary, add a description of why the outcome is bad and what would you expect 'good' to look like.
```

- Issue titles should _not_ contain any prefixes or namespaces, instead this should be part of
  the problem statement. I.e. "Write a document explaining the topic" rather than "[doc/explanation]
  Topic".

### Issue Estimation

- Issues should be estimated using the "Original time estimate"-field in Jira.
- Issues expected to take less than a full day of work (8 hours) should be estimated in full hours.

> - Issues expected to take longer than a full day of work (8 hours) should be estimated in whole days.

Reviewer: ```suggestion
- Issues expected to take longer than a full day of work (8 hours) should be estimated in multiples of 8 hours (8 hours=1 day).
```

### Pull Requests

- Pull request titles should be following the conventional commit format, which you can read
  [here](https://www.conventionalcommits.org/en/v1.0.0/). In essence,

  > <type>[(<scope>)]: description

  for example:

>   > bug(charm-tracing): call the `setup` function whenever something is being set up

Author: ```suggestion
  > fix(charm-tracing): call the `setup` function whenever something is being set up
```

>   > bug(charm-tracing): call the `setup` function whenever something is being set up

Author: > this does not look like conventional commits.

If you are thinking of the `bug` type, you are right - changed it to `fix`. With regards to the actual format: it is.

> Also, conventional commits are about commits.

they are, but since we are squashing our commits, the pr title is what ends up being the commit message of the merged commit. While it is possible to edit it manually at the merge step, it's definitely easier if we just start doing it in the PR title so it's right regardless of who pushes the merge button. 

> [the square bracket examples at the end]

There are no square brackets in conventional commit types.

# Creating issues in GitHub and Jira

**Date:** 2024-12-03<br/>
**Authors:** @simskij

## Context and Problem Statement

As the team have grown, we've ended up diverging in our practices on how to create issues in our
repositories. To be able to work efficiently and consistently with our issues, we need to follow some
common practices, so that the workflow remains the same across issues.

## Decision

These rules only apply to tickets opened by us. If someone else opens the ticket, we should
rephrase it as part of the triage.

### Issue Titles

- Issues are always to be created in GitHub, unless confidentiality requirements dictate something
  else. However, this is almost never the case. If you are unsure - ask @simskij or @sed-i.
- Issues of the story/enhancement type should have a title that describes the work to be done in
  the _imperative modus_, i.e "Add a bip to the bop", rather than "We should add a bip to the bop"
- Issues of the bug type should have a title that describes the bug occuring, written in

>   this format: "When <condition> then <outcome>".

Reviewer 2: It's surpisingly hard to describe complex scenarios that way. Just started writing a bug named 

"When charm is related to grafana-agent without a connected tracing backend then cos-agent exposes an inaccessible tracing URL" which doesn't sound easy to say in a status update. I think it makes sense as a guideline, but not necessarily a rule.

> - Issues of the bug type should have a title that describes the bug occuring, written in
>   this format: "When <condition> then <outcome>".

Author: ```suggestion
- Issues of the bug type should have a title that describes the bug occuring, preferably
  written in this format: "When <condition> then <outcome>".
```

- Issue titles should _not_ contain any prefixes or namespaces, instead this should be part of
  the problem statement. I.e. "Write a document explaining the topic" rather than "[doc/explanation]
  Topic".

### Issue Estimation

- Issues should be estimated using the "Original time estimate"-field in Jira.
- Issues expected to take less than a full day of work (8 hours) should be estimated in full hours.
- Issues expected to take longer than a full day of work (8 hours) should be estimated in whole days.

### Pull Requests

- Pull request titles should be following the conventional commit format, which you can read
  [here](https://www.conventionalcommits.org/en/v1.0.0/). In essence,

>   > <type>[(<scope>)]: description
> 
>   for example:
> 
>   > fix(charm-tracing): call the `setup` function whenever something is being set up

Author: ```suggestion
  ```
  <type>[(<scope>)]: description
   ```

  for example:

  ```
  fix(charm-tracing): call the `setup` function whenever something is being set up
  ```
```

>   The valid prefixes are:

Author: ```suggestion
  The valid types are:
```
